library/alloc/src/boxed.rs

```diff
@@ -1105,8 +1142,9 @@ impl<T: ?Sized, A: Allocator> Box<T, A> {
     /// assert_eq!(*static_ref, [4, 2, 3]);
     /// ```
     #[stable(feature = "box_leak", since = "1.26.0")]
+    #[rustc_const_unstable(feature = "const_box", issue = "none")]
     #[inline]
-    pub fn leak<'a>(b: Self) -> &'a mut T
+    pub const fn leak<'a>(b: Self) -> &'a mut T
```

Comment:
This should be illegal because it points to static memory.

Reply:
This is why we need the magic first... Or we add a `const_leak` that only returns an immutable reference.

Reply:
Well, it is only a problem with ConstAlloc. Using an arena should be fine, and that is the only thing it really enables

Reply:
I think ultimately this PR should be postponed. If this PR is to be landed right now, I would tell people not to use it because it is very easy to write unsound code that uses these functions.

Reply:
I thought that mutable references or pointers can't be leaked ouf of const context (compile error), is this assumption wrong?

Reply:
How is a box different to a newtype `PhantomData<T>`? I'd assume not very different.

```rust
use std::cell::Cell;
use std::marker::PhantomData;

pub struct MyBox<T>(PhantomData<T>);

pub const A: &'static MyBox<Cell<i32>> = &MyBox(PhantomData); 
```

This already compiles today. I'd like to retract my comment that it is easy to write unsound code, though.

([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4a69cfa73a73a299fbcf1ae231084370))

Reply:
> But then we can't add `ConstAlloc` without making this non-const first.

This is just a convenience function. Using `ConstAlloc` people can build the very same abstraction in user code. The problem lies with `ConstAlloc` from my perspective.

I do recognize the point about "don't use this"... With arenas it's fine, and that is the only thing people can build without resorting to the const_allocate intrinsic.

I'm very much for not stabilizing any of this before we figure out the const  heap story. Do you think we should hold back making these functions and types unstably const?

Reply:
I don't know if there are any policy for adding unstable functionality that could cause UB. If you're fine with this, then it's okay I guess.

Reply:
Unstable features should be sound -- or if it is unclear how to achieve that, they should trigger the `incomplete_features` lint.

Reply:
My thoughts exactly. I don't think there is a mechanism for declaring incomplete *library* features, though.

Reply:
Hold up. How is it unsound right now? It would only be unsond with a const allocator that uses unstable features, right? It is perfectly safe with an arena allocator, as values allocated with that can never leave the const item

Reply:
You can cause UB with an allocator using `const_allocate` and end up having a static reference to types containing interior mutability in safe rust with this change I think. And that is unsound

Reply:
Yea, but this PR doesn't change that. You can do this right now. That is more an argument to remove `const_allocate` to me

Reply:
Sorry for the confusion. I think the solution is a follow-up PR that documents the safety of the `const_allocate` intrinsic. Then we'd categorize the UB that users can write as "misuse of the unsafe `const_allocate` function".